### PR TITLE
Feature/add customer layers

### PR DESCRIPTION
**Pull request by BartoliniAndBorderCollies**
#### Issue
https://trello.com/c/uGrX7YJv

### Description
This branch adds controller, service layers for customer operations. From now on it is possible to add, delete, update and read customers.

### Motivation
It is a basic feature to be able to handle customers in application. Through "customers" I mean all types of users in app.

### Changes introduced by this pull request

- add CustomerDTOs classes
- add CustomerController, CustomerService classes with CRUD methods
- add jakarta validations to Customer and CustomerDTORequest classes
- add hibernate-validator dependency
- add validation logic to create() in CustomerService class
- add handleConstraintViolationException() in GlobalCustomExceptionHandler
- add security paths for customer actions in SecurityConfig class
- add Junit tests
- add integration testing
- add equals() hashCode() to Authority class

### Testing
- [x]  all Junit and integration tests are functional and operational
- [x] 100% methods and 100% lines covered at CustomerController
- [x] 100% methods and 95% lines covered at CustomerService
- [x] 100% methods and 100% lines covered at CustomerDetailsService

### CODE REVIEW

File: src/integration_test/resources/application-test.properties
Comment (JasinskiSz) [inline src/integration_test/resources/application-test.properties:1]: Same case as in the PR #4

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: Why would you like to return a password in response? Is this necessary in your case?

**Comment by JasinskiSz** [comment]
There are changes that are coming in #4 so this PR shouldn't be marked as "code review ready". In cases like that this should be placed to check after PR #4 is merged or PR #5 should only introduce new changes without those from PR #4. I believe you'll merge PR #4 soon, so then I'll review this PR in full :) I've already checked some of the changes, but I don't want to miss anything, so I'll wait for merge of previous PR.

**Comment by BartoliniAndBorderCollies** [comment]
okay I have merged a branch to develop, later on I merge develop to this branch.
Which means this branch is ready for review!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java]:
```suggestion
            "Password must have at least 5 characters, one digit, one lowercase letter, one uppercase letter, and a special character")
```

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java]: I'd sort those fields more accurately to their usage. By that I mean to place `private String emailAddress;` and `private Set<Authority> authorities;` above all the Boolean fields that you have commented as hardcoded. This will make the code more readable. It'd be very sexy to place distinct those hardcoded fields from all the others by making some comments in the code. This seriously improves code readability.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/configuration/SecurityConfig.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/configuration/SecurityConfig.java]: This comment made me believe that below are patterns **reachable** for a customer and not the patterns that **include** "customer" in uri path :) My suggestion is to clear out confusion that could make by changing the comment to be more descriptive, but also short.

To code below I have no other comments, it's all looking good.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java]: Would be nicer to create a message that would be more meaningful, more understandable. The wording of that message means that email **must be** the example you provided.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/CustomerController.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/CustomerController.java]: Should be final
```suggestion
    private final CustomerService customerService;
```

File: src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/advice/GlobalCustomExceptionHandler.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/advice/GlobalCustomExceptionHandler.java]: Instead of 400 code, the 422 could be used:
https://stackoverflow.com/questions/16133923/400-vs-422-response-to-post-of-data

At least that what I'd do and imo it's a case here, too.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java]: Final fields should be here.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java]: There is a lot happening for this one `create` method. It'd be nice to split this method into separate, private methods or to another service for validation only. Right now this is not exactly following "**S**OLID" rule.

The method right now Is responsible for:
1) saving entity
2) mapping entity – which is send to ModelMapper (that's cool)
3) validating CustomerDTORequest
4) fetching authorities from db

The last two parts are pretty big and to follow good practices it's advisable to give them their own space instead of squishing it all in `create` method.

**Review by JasinskiSz — COMMENTED** [review]
Take a look at my findings in the code :)

File: src/integration_test/resources/application-test.properties
Comment (BartoliniAndBorderCollies) [reply src/integration_test/resources/application-test.properties:1]: done!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: no, you are right! especially that password is being hashed!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java]: great!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java]: allright! I have just moved String emailAddress and Set<Authority> authorities to the above but about the second part I must say that I have put those fields to avoid hardcoding, so they are not harcoded fields. 
They would be but in Customer class I set they return real values:

image

File: src/main/java/com/example/Restaurant_Management_System_REST_API/configuration/SecurityConfig.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/configuration/SecurityConfig.java]: allright! I hope now its more descriptive, thanks for that!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/advice/GlobalCustomExceptionHandler.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/rest/controller/advice/GlobalCustomExceptionHandler.java]: OK, I see the point!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: Oh now I remember, if I delete password field from here then I'm not able to update password in update():

image

File: src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java]: right!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java]: done!

**Comment by BartoliniAndBorderCollies** [comment]
thanks for your inquiries! ready for re-review!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: Yep, I see. But you can easily overcome this issue. Do you have any ideas how this could be done?

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java]: I didn't mean to move anything to other classes or I did not said nothing about changing the hardcoded part. I meant to rearrange lines of code (the fields part) in the class where I put the comment.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java]: Is there a reason behind `static` here? sorry if I can't see obvious here :)

**Review by JasinskiSz — COMMENTED** [review]
SecurityConfig imo still doesn't have descriptive enough comments that leads me to the same conclusion as before.

Other than that you've implemented other corrections in a very good fashion :)

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: I have some idea, what do you think about this?

image

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java]: I haven't moved anything to other class ;) I just move them above. I haven't changed the hardcoded part neither! :)

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: and also I would need to change integration tests because now I check if the password provided is the same as it was returned. for example here:

image

File: src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java]: the reason was I didnt notice while it was extracted. Static is not needed according to me

**Comment by BartoliniAndBorderCollies** [comment]
ready for re-review! and I left some answers to your concerns as well. please answer

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: and this is how could I remake the integration test to still check if password is correct. tell me what you think about this:

image

Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: or I could resign totally from checking the passwords -> I dont know how it is in real commercial projects..

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTORequest.java]: I believe we didn't understood each other then :)

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: What you had in the beginning was okay. The problem was only to return password as a response, especially in the encoded form as this makes whole password security vulnerable.
I'd suggest simply omitting password in response :)

**Review by JasinskiSz — APPROVED** [review]
Looks like most/all of the major problems are gone. Great, Bartek!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/DTO/CustomerDTOs/CustomerDTOResponse.java]: this is still unknown to me @JasinskiSz 
1. is update() now correct? look above - I added a screenshot
2. for the integration tests I could remove checking password at all or I could remake these methods like on example above. Please take a look and let me know!

**Pull request merged**
